pkg/kubelet/cm/pod_container_manager_linux.go

```diff
@@ -108,10 +114,101 @@ func (m *podContainerManagerImpl) GetPodContainerName(pod *api.Pod) string {
 	return path.Join(parentContainer, podContainer)
 }
 
+// Scans through all the subsystems pod cgroups directory.
+// Will also scan through the container cgroups that still exist under
+// the pod cgroup and haven't been Garbage Collected yet
+func (m *podContainerManagerImpl) getPIDsToKill(podCgroup string) []int {
+	// Get a list of processes that we need to kill
+	pidsToKill := sets.NewInt()
+	var pids []int
+	for _, val := range m.subsystems.MountPoints {
+		dir := path.Join(val, podCgroup)
+		_, err := os.Stat(dir)
+		if os.IsNotExist(err) {
+			// The subsystem pod cgroup is already deleted
+			// do nothing, continue
+			continue
+		}
+		// Get a list of pids that are still charged to the pod's cgroup
+		pids, err = getCgroupProcs(dir)
+		if err != nil {
+			continue
+		}
+		pidsToKill.Insert(pids...)
+
+		// WalkFunc which is called for each file and directory in the pod cgroup dir
+		visitor := func(path string, info os.FileInfo, err error) error {
+			if !info.IsDir() {
+				return nil
+			}
+			pids, err = getCgroupProcs(path)
+			if err != nil {
+				return err
+			}
+			pidsToKill.Insert(pids...)
+			return nil
+		}
+		// Walk through the pod cgroup directory to check if
+		// container cgroups haven't been GCed yet. Get attached processes to
+		// all such unwanted containers under the pod cgroup
+		err = filepath.Walk(dir, visitor)
+	}
+	return pidsToKill.List()
+}
+
+// Scan through the whole cgroup directory and kill all processes either
+// attached to the pod cgroup or to a container cgroup under the pod cgroup
+func (m *podContainerManagerImpl) tryKillingCgroupProcesses(podCgroup string) error {
+	pidsToKill := m.getPIDsToKill(podCgroup)
+	// No pids charged to the terminated pod cgroup return
+	if len(pidsToKill) == 0 {
+		return nil
+	}
+
+	var errlist []error
+	// os.Kill often errors out,
+	// We try killing all the pids multiple times
+	for i := 0; i < 5; i++ {
+		if i != 0 {
+			glog.V(3).Infof("Attempt %v failed to kill all unwanted process. Retyring", i)
```

Comment:
As of now, this PR is not even ready since the pod level cgroups are
causing tests to flake randomly. Given the experimental nature of it, I
don't see why we would restrict it in v1.4 if this PR were to be
hypothetically ready to be merged.

On Tue, Aug 23, 2016 at 12:08 PM, Derek Carr <notifications@github.com>
wrote:

> In pkg/kubelet/cm/pod_container_manager_linux.go
> <https://github.com/kubernetes/kubernetes/pull/29049#discussion_r75931112>
> :
>
> > +
> > +// Scan through the whole cgroup directory and kill all processes either
> > +// attached to the pod cgroup or to a container cgroup under the pod cgroup
> > +func (m *podContainerManagerImpl) tryKillingCgroupProcesses(podCgroup string) error {
> > +	pidsToKill := m.getPIDsToKill(podCgroup)
> > +	// No pids charged to the terminated pod cgroup return
> > +	if len(pidsToKill) == 0 {
> > +		return nil
> > +	}
> > +
> > +	var errlist []error
> > +	// os.Kill often errors out,
> > +	// We try killing all the pids multiple times
> > +	for i := 0; i < 5; i++ {
> > +		if i != 0 {
> > +			glog.V(3).Infof("Attempt %v failed to kill all unwanted process. Retyring", i)
>
> throughout this pr, can we prefix any logging from this module with
> something common?
>
> it will make log analysis around pod level cgroups easier to grep against
> when handling support problems.
>
> something like: "pod container manager: "
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/29049/files/64c3e8845edec7722bc20b9a94fb32d9253841bc#r75931112>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGvIKDteCvog1R2JOy58wyocJ5hZ2v0Aks5qi0UggaJpZM4JN7WQ>
> .
>